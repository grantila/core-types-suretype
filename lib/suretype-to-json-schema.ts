import type { ConversionResult } from 'core-types'
import { decorateSchema } from 'core-types-json-schema'
import { basename } from 'path'

import type { SuretypeToJsonSchemaOptions } from './types.js'
import { extractJsonSchema } from './read-exports/index.js'
import { catchError } from './async-error.js'


export async function convertSuretypeToJsonSchema(
	filename: string,
	options: SuretypeToJsonSchemaOptions = { }
)
: Promise< ConversionResult< { } > >
{
	const {
		warn = console.warn.bind( console ),
		refMethod = 'provided',
		nameConflict = 'error',
	} = options ?? { };

	const { result, error } = await catchError(
		( ) => extractJsonSchema(
			filename,
			refMethod,
			nameConflict === 'rename' ? 'rename' : 'error'
		)
	);

	if ( error )
	{
		if ( error.name?.startsWith( 'Duplicate' ) && nameConflict === 'warn' )
			warn( error.message, error );
		else
			throw error;
	}

	if ( !result )
		return { data: { }, convertedTypes: [ ], notConvertedTypes: [ ] };

	decorateSchema(
		result.jsonSchema,
		{ sourceFilename: basename( filename ), ...options },
		'core-types-suretype',
		'https://github.com/grantila/core-types-suretype',
	);

	return {
		data: result.jsonSchema,
		convertedTypes: Object.keys( result.jsonSchema.definitions! ),
		notConvertedTypes: [ ],
	};
}
