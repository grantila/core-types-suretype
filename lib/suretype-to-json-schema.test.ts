import path from 'path'
import { fileURLToPath } from 'url'

import { convertSuretypeToJsonSchema } from './suretype-to-json-schema.js'


const __dirname = path.dirname( fileURLToPath( import.meta.url ) );

describe( "convertSuretypeToJsonSchema", ( ) =>
{
	it( "should convert types with dependencies", async ( ) =>
	{
		const file = path.join( __dirname, 'test', 'validator.js' );
		const res = await convertSuretypeToJsonSchema( file );
		const { convertedTypes, notConvertedTypes, data } = res;

		expect( convertedTypes.sort( ) )
			.toStrictEqual( [ 'ChatLine', 'User' ] );
		expect( notConvertedTypes ).toStrictEqual( [ ] );

		expect( data ).toMatchSnapshot( );
	} );
} );
