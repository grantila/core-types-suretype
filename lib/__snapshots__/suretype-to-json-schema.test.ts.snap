// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertSuretypeToJsonSchema should convert types with dependencies 1`] = `
{
  "$comment": "Generated from validator.js by core-types-suretype (https://github.com/grantila/core-types-suretype)",
  "definitions": {
    "ChatLine": {
      "properties": {
        "line": {
          "type": "string",
        },
        "user": {
          "$ref": "#/definitions/User",
        },
      },
      "required": [
        "line",
      ],
      "title": "A chat line",
      "type": "object",
    },
    "User": {
      "description": "A user representation",
      "examples": [
        "{ firstName: "joe", id: "12345678" }",
      ],
      "properties": {
        "firstName": {
          "type": "string",
        },
        "id": {
          "minLength": 8,
          "type": "string",
        },
      },
      "required": [
        "firstName",
        "id",
      ],
      "title": "The user type",
      "type": "object",
    },
  },
}
`;
